Updated IdentityProviders to treat 'origin' as an identifier

tvs opened the pull request:
We were previously treating 'origin' as a sort of hardwired type-identifier, when it should have been treated as a unique name. While 'uaa' and 'LDAP' had special meanings, they were still not 'types'.

tvs committed:
Added general-purpose SCIM management through VMidentity.

Some concepts are not yet mapped such as GroupExternalMembers.
Additionally, ScimGroupMember is not yet defined as some providers do
not allow filtering on the 'member' field (yet).

tvs committed:
Added client query functionality

Implemented preliminary sorting functionality

tvs committed:
Updated IdentityProviders to treat 'origin' as a name

Previously it was being treated as an identity provider type rather than
as an identifier.

tvs committed:
Updated leftover origin references in Vmidentity code

tvs committed:
Fixed possible NPE

tvs committed:
Treat origin as a domain name not a provider name

tvs committed:
Removed solution users from membership queries

Fixed a few typos in log files

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/authentication/manager/VmidentityAuthenticationManager.java:
```diff
@@ -137,7 +134,7 @@ public Authentication authenticate(Authentication req) throws AuthenticationExce
                         upn,
                         upn,
                         upn,
-                        userId.getDomain().equalsIgnoreCase(systemDomain) ? OriginKeys.UAA : OriginKeys.LDAP,
```
for future, it would be really nice if we could separate a logical change vs reformat/var rename change.
it is really hard to review when this is all munged together. thanks!

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/provider/VmidentityIdentityProviderProvisioning.java:
```diff
@@ -80,8 +81,8 @@ public IdentityProvider retrieve(String id) {
             // TODO: active only - we only have active right now...
-            String tenant = VmidentityUtils.getTenantName(zoneId, this._idmClient.getSystemTenant());
-            Collection<IIdentityStoreData> providers = this._idmClient.getProviders(
+            String tenant = VmidentityUtils.getTenantName(zoneId, this.client.getSystemTenant());
```
i'm not sure this approach will work. we seem to map proider id, to idm's provider name. Yet we put domain name for user origin. i think this won;t wotk. provider name != [provider domain, and more over some providers have multiple domains. why do we not unify on provider is = idm's provider anme ? this seems easier to handle ....

mpotapova-vmw reviewed (commented):
thanks. couple questions below.

tvs committed:
Clean up some of the sorting code

Used less verbose constructions in place of the lambdas where possible.

tvs committed:
Updated a few of the sorting calls

Cleaned up and commented a few of the remaining areas

tvs committed:
Updated a few of the sorting calls

Cleaned up and commented a few of the remaining areas

tvs committed:
Cleaned up imports on CompareUtils

tvs committed:
Merge branch 'vmidentity-scim' into vmidentity-idp

tvs committed:
Use provider name for an origin rather than domain

Updated authentication, users and group members to use a provider name
for the origin rather than a domain.

Group members can originate from other domains (due to FSPs) so we need
to check on each member whether the origin should differ.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/user/VmidentityUserDatabase.java:
```diff
@@ -96,7 +96,7 @@ public UaaUser retrieveUserById(String id) throws UsernameNotFoundException {
 
             String[] parts = id.split("@");
-            String origin = (systemDomain.equalsIgnoreCase(parts[1])) ? OriginKeys.UAA : OriginKeys.LDAP;
+            String origin = (systemDomain.equalsIgnoreCase(parts[1])) ? OriginKeys.UAA : parts[1];
```
this is wrong, we need origin, now it seems to be domain

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/util/VmidentityUtils.java:
```diff
@@ -35,15 +38,31 @@
 
     public static String getSystemDomain(String tenant, CasIdmClient idmClient) throws Exception {
         Collection<IIdentityStoreData> provider = idmClient.getProviders(tenant, EnumSet.of(DomainType.SYSTEM_DOMAIN));
-        // todo: should get security domain name, not provider name.
```
why is this ok to leave the name here. name is not the domain name....

mpotapova-vmw replied:
let's keep this todo (technically we should return domain name, right now for system domain they are always the same)....
and submit....

mpotapova-vmw reviewed (commented):
overall looks good. couple questions below. thanks.

tvs committed:
Fixed issue with origins and origin-checking in VmidentityUserDatabase

tvs committed:
Merge branch 'vmidentity' into vmidentity-idp
